README.md

Do it all! in Golang
====================
[![CircleCI Build Status](https://circleci.com/gh/pathaugen/doitall-golang.svg?style=svg)](https://circleci.com/gh/pathaugen/doitall-golang)
[![TravisCI Build Status](https://travis-ci.com/pathaugen/doitall-golang.svg?branch=master)](https://travis-ci.com/pathaugen/doitall-golang)
[![Build status](https://ci.appveyor.com/api/projects/status/2hm4g4of8d2k5of1/branch/master?svg=true)](https://ci.appveyor.com/project/PatrickHaugen/doitall-golang/branch/master)

[![Coverage Status](https://coveralls.io/repos/github/pathaugen/doitall-golang/badge.svg?branch=master)](https://coveralls.io/github/pathaugen/doitall-golang?branch=master)
[![Maintainability](https://api.codeclimate.com/v1/badges/4fa3e21b6359a2da09ce/maintainability)](https://codeclimate.com/github/pathaugen/doitall-golang/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/4fa3e21b6359a2da09ce/test_coverage)](https://codeclimate.com/github/pathaugen/doitall-golang/test_coverage)

Description
-----------

Simple project framework to provide code and a baseline to do just about anything in Go.

Quick Start
------------

1. Machine Setup
2. Compile and Run the Application
3. Operations Available:
   * Math Operations:
     * Addition
     * Subtraction
     * Multiplication
     * Division
   * API Operations:
     * Define API Endpoint

---------- Machine Setup ----------

1. Install Go:
   * https://golang.org/dl/
   * Recommend C:\Go on Windows
   * Test Go Installation:
     * > go version
2. Setup Go:
   * Create File Structure:
     * C:\go-work\bin
     * C:\go-work\pkg
     * C:\go-work\src
   * Check Environment Variables:
     * GOROOT: C:\Go
     * GOPATH: C:\go-work
3. Setup Git:
   * Install Git:
   * x
x. Test Code:
   * > go test -v
   * xxx
x. Check Documentation:
   * > godoc -http=:6060
   * http://localhost:6060/pkg/
     * All code in $GOPATH (which includes stdlib) appears in web interface.

---------- Compile and Run the Application ----------

$ go build
  Compiles the application.

$ doitall-golang
  Runs the application.

$ doitall-golang -changelog
  View the changelog.

---------- Math Operations ----------

  Addition:
  $ doitall-golang -add X Y
    Adds the variables X and Y

  Subtraction:
  $ doitall-golang -sub X Y
    Adds the variables X and Y

  Multiplication:
  $ doitall-golang -mult X Y
    Adds the variables X and Y

  Division:
  $ doitall-golang -div X Y
    Adds the variables X and Y

---------- API Operations ----------

  Define API Endpoint:
  $ doitall-golang -api X
    Specifies X as the API endpoint to utilize for API operations.

Features
--------

(developing)

External Libraries
------------------

(developing)


CHANGELOG.md

# Project Change Log

**ATTN**: This project uses [semantic versioning](http://semver.org/).

***

## 1.0.0 - 2019-01-13

### Added
Version 0.0.2
* x

### Changed

### Fixed

***

## 1.0.0 - 2019-01-12

### Added
Version 0.0.1
* Splash ASCII
* Basic functions
* Reading flags from command line

### Changed

### Fixed

***


functions_test.go

package main

import "testing"


// ========== ========== ========== ========== ==========
func TestAdd(t *testing.T) {
  t.Log("Adding the numbers 12 and 28... (expected result: 40)")

  result := Add(12,28)

  if result != 40 {
    t.Errorf("Expected result of 40, but it was %d instead.", result)
  }

}
// ========== ========== ========== ========== ==========


// ========== ========== ========== ========== ==========
func TestAddArray(t *testing.T) {
  t.Log("Using an array and SumArray(), Add() all values in a loop... (expected result: 75)")

  _, result := SumArray( []int{ 10, 11, 12, 13, 14, 15 } )

  if result != 75 {
    t.Errorf("Expected result of 75, but it was %d instead.", result)
  }

}
// ========== ========== ========== ========== ==========


functions.go

package main

import (
	//"fmt"

	"strconv" // https://golang.org/pkg/strconv/
)


// ========== START: Add(x,y) Addition ========== ========== ========== ==========
func Add(x int, y int) int {
	return x + y
}
// ========== END: Add(x,y) Addition ========== ========== ========== ==========


// Subtraction

// Multiplication

// Division


// ========== START: SumArray( array ) ========== ========== ========== ==========
// Take an array input and add all values together
// Can break array in half and utilize goroutines and a channel to split the work: https://tour.golang.org/concurrency/2
func SumArray( inputArray []int ) (string, int) {
  outputString := ``
  resultInt := 0


  // ==========
  //for i := 0; i < 10; i++ {
    //outputString += "  interval: " + strconv.Itoa(i) + "\n" // Itoa is shorthand for FormatInt(int64(i), 10)
    // //sum += i
    //resultInt = Add( resultInt, i)
    //outputString += "    sum: " + strconv.Itoa(resultInt) + "\n"
  //}
  // ==========

  // ==========
  for i := 0; i < len(inputArray); i++ {
    outputString += "  interval: " + strconv.Itoa(i) + "\n" // Itoa is shorthand for FormatInt(int64(i), 10)
    resultInt = Add( resultInt, inputArray[i] )
    outputString += "    sum: " + strconv.Itoa(resultInt) + "\n"
  }
  // ==========

  outputString += "  Final sum: " + strconv.Itoa(resultInt)

	return outputString, resultInt
}
// ========== END: SumArray( array ) ========== ========== ========== ==========


main.go

package main

import (
	"fmt"

	"net/http"

	//"strconv" // https://golang.org/pkg/strconv/
	"strings"

	"bufio"
	"os"
	"os/exec"

	"time"
	//"log"
)



// NOTES:
// Golang Slices are basically Arrays
// Golang Maps are basically Dicts

// TODO:
// - reorder array alphabetically
// - pull json and parse out, putting into struct first and utilizing there


// Pull changelog.md and display

// Pull readme.md as documenation and display


// ========== START: Golang Console Colors ========== ========== ========== ==========
// Example: fmt.Print( cRed + "HelloWorld" + cClr )
var cClr				= "\u001b[0m"

var cBold				= "\u001b[1m"

var cBlack			= "\u001b[30m"
var cRed				= "\u001b[31m"
var cGreen			= "\u001b[32m"
var cYellow			= "\u001b[33m"
var cBlue				= "\u001b[34m"
var cMagenta		= "\u001b[35m"
var cCyan				= "\u001b[36m"
var cWhite			= "\u001b[37m"

var cBlackBG		= "\u001b[40m"
var cRedBG			= "\u001b[41m"
var cGreenBG		= "\u001b[42m"
var cYellowBG		= "\u001b[43m"
var cBlueBG			= "\u001b[44m"
var cMagentaBG	= "\u001b[45m"
var cCyanBG			= "\u001b[46m"
var cWhiteBG		= "\u001b[47m"
// ========== END: Golang Console Colors ========== ========== ========== ==========


var appinfo = `
  ` + cBlue + `=======================================` + cBold + cCyan + `
       _           _  _             _  _
      | |         (_)| |           | || |
    __| |  ___     _ | |_     __ _ | || |
   / _`+"`"+` | / _ \   | || __|   / _`+"`"+` || || |
  | (_| || (_) |  | || |_   | (_| || || |
   \__,_| \___/   |_| \__|   \__,_||_||_|` + cClr + `
                             ` + cCyan + `..in Golang!` + cClr + `
  ` + cBlue + `=======================================` + cClr + `
`


// > go build&doitall-golang.exe
func main() {


	// ========== ========== ========== ========== ==========
	// Declare and Server in one go
	webserverRunning := false
	http.HandleFunc( "/", func(w http.ResponseWriter, r *http.Request ) {
		//fmt.Fprintf(w, "Hello, %q", html.EscapeString(r.URL.Path))
		//fmt.Fprintf( w, "This is / of the Webserver" )
		fmt.Fprintf( w, htmlTemplate )
	})
	http.HandleFunc("/api", handlerApi)
	// ========== ========== ========== ========== ==========


	// https://stackoverflow.com/questions/22891644/how-can-i-clear-the-terminal-screen-in-go#22896706
	cmd := exec.Command("cmd", "/c", "cls")
	cmd.Stdout = os.Stdout
	cmd.Run()

	breakspace := ("\n")
	//breakline := ( breakspace + cBlue + "  ========== ========== ========== ========== ==========" + cClr + breakspace )
	breakline := ( breakspace + cBlue + "  =======================================" + cClr + breakspace )

	// Start Reading Input
	scanner := bufio.NewScanner(os.Stdin)
	var inputText string

	for ( inputText != "q" ) {  // break the loop if inputText == "q"

		cmd := exec.Command("cmd", "/c", "cls")
		cmd.Stdout = os.Stdout
		cmd.Run()

		// ========== ========== ========== ========== ==========
		// Take Command Line Arguments
		// fmt.Print("Take Command Line Arguments:" + breakspace)
		// ========== ========== ========== ========== ==========

		fmt.Print(
			appinfo + breakspace +
			"  Do it all in Golang!" + breakspace +
			"  https://github.com/pathaugen/doitall-golang" + breakspace )

		//var b []byte = make([]byte, 1)
		//for {
			//os.Stdin.Read(b)
			//fmt.Println("I got the byte", b, "("+string(b)+")")
		//}

		if ( inputText == "s" ) {

			fmt.Print( breakline )

			// ========== START: Golang Internal Webserver ========== ========== ========== ==========
			// Channels: https://tour.golang.org/concurrency/2
			fmt.Print(
				breakspace + cBold + cCyan + "  Start/Stop Internal Webserver:" + cClr + breakspace )

			fmt.Print( breakspace )

			//webserverChannel := make(chan bool)
			//webserverChannel <- false

			fmt.Print( cBold + cGreen + "  Starting Internal Webserver on localhost port :9090" + cClr )
			go func() {
				//fmt.Print( "Starting Webserver on localhost port :9090" )

				//webserverChannel <- true

				http.ListenAndServe(":9090", nil)
				//time.Sleep(time.Second * 3)

				//webserverChannel <- false
			}()
			webserverRunning = true

			// log.Fatal( http.ListenAndServe(":9090", nil) )
			//err := http.ListenAndServe(":9090", nil) // set listen port
			//if err != nil {
				//log.Fatal("ListenAndServe: ", err)
			//}

			//fmt.Print( <-webserverChannel )
			//for i := 0; i < len(webserverChannel); i++ {
				//fmt.Print( <-webserverChannel )
			//}

			// ========== END: Golang Internal Webserver ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "t" ) {

			fmt.Print( breakline )

			// ========== ========== ========== ========== ==========
			// Run Tests
			fmt.Print(
				breakspace + cBold + cCyan + "  Golang Tests:" + cClr + breakspace )

			fmt.Print( breakspace )

			testOutput, _ := exec.Command("cmd", "/c", "go test -v").Output()
			//cmd.Stdout = os.Stdout
			//cmd.Run()

			// Add Colors to testOutput
			testOutputString := string(testOutput)
			testOutputString = strings.Replace( testOutputString, "RUN", cCyan + "RUN" + cClr, -1 )
			testOutputString = strings.Replace( testOutputString, "PASS", cGreen + "PASS" + cClr, -1 )
			testOutputString = strings.Replace( testOutputString, "FAIL", cRed + "FAIL" + cClr, -1 )

			fmt.Print( testOutputString )

			// ========== ========== ========== ========== ==========

			//fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "0" ) {

			fmt.Print( breakline )

			// ========== ========== ========== ========== ==========
			// Color Test

			fmt.Print(
				breakspace + cBold + cCyan + "  Golang Console Color Testing:" + cClr + breakspace +

				breakspace + "  Basic Text:" + cClr + cBold + "    Bold Text:" + cClr + "     Basic Background:" + cClr +
				breakspace +
				breakspace + cBlack + "  Black" + cClr + "          " + cBold + cBlack + "Black" + cClr + "          " + cBlackBG + "Black" + cClr +
				breakspace + cRed + "  Red" + cClr + "            " + cBold + cRed + "Red" + cClr + "            " + cRedBG + "Red" + cClr +
				breakspace + cGreen + "  Green" + cClr + "          " + cBold + cGreen + "Green" + cClr + "          " + cGreenBG + "Green" + cClr +
				breakspace + cYellow + "  Yellow" + cClr + "         " + cBold + cYellow + "Yellow" + cClr + "         " + cYellowBG + "Yellow" + cClr +
				breakspace + cBlue + "  Blue" + cClr + "           " + cBold + cBlue + "Blue" + cClr + "           " + cBlueBG + "Blue" + cClr +
				breakspace + cMagenta + "  Magenta" + cClr + "        " + cBold + cMagenta + "Magenta" + cClr + "        " + cMagentaBG + "Magenta" + cClr +
				breakspace + cCyan + "  Cyan" + cClr + "           " + cBold + cCyan + "Cyan" + cClr + "           " + cCyanBG + "Cyan" + cClr +
				breakspace + cWhite + "  White" + cClr + "          " + cBold + cWhite + "White" + cClr + "          " + cWhiteBG + "White" + cClr )
			// ========== ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "1" ) {

			fmt.Print( breakline )

			// ========== ========== ========== ========== ==========
			// Simple Addition: Add(x,y)
			fmt.Print(
				breakspace + cBold + cCyan + "  Simple Addition: Add(x,y)" + cClr + breakspace )

			fmt.Print( breakspace )

			//fmt.Print("Calling function add with 42, 13: " + add(42, 13))

			fmt.Print( Add(42, 13) )
			// ========== ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "2" ) {

			fmt.Print( breakline )

			// ========== ========== ========== ========== ==========
			// Add Array Values in Loop: SumArray( array )
			fmt.Print(
				breakspace + cBold + cCyan + "  Add Array Values in Loop: SumArray( array )" + cClr + breakspace )

			fmt.Print( breakspace )

			// Define an Array
			inputArray := []int{ 10, 11, 12, 13, 14, 15 }
			funcOutput, _ := SumArray( inputArray )
			fmt.Print( funcOutput )

			// ========== ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "3" ) {

			fmt.Print( breakline )

			// ========== ========== ========== ========== ==========
			// Load JSON into Array
			fmt.Print(
				breakspace + cBold + cCyan + "  Load JSON into Array:" + cClr + breakspace )
			// ========== ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "4" ) {

			// ========== ========== ========== ========== ==========
			// Reorder Array Alphabetically
			fmt.Print(
				breakspace + cBold + cCyan + "  Reorder Array Alphabetically:" + cClr + breakspace )
			// ========== ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "5" ) {

			// ========== ========== ========== ========== ==========
			fmt.Print(
				breakspace + cBold + cCyan + "  Goroutines Communicating via Channels" + cClr + breakspace )

			goroutinecheck := make(chan int)
			go func() {
				time.Sleep(time.Second * 3)
				goroutinecheck <- 1
			}()
			go func() {
				time.Sleep(time.Second * 2)
				goroutinecheck <- 2
			}()
			go func() {
				time.Sleep(time.Second * 1)
				goroutinecheck <- 3
			}()
			for i := 0; i < 3; i++ {
				fmt.Print( <-goroutinecheck )
			}

			// ========== ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else if ( inputText == "6" ) {

			// ========== ========== ========== ========== ==========
			fmt.Print(
				breakspace + cBold + cCyan + "  OPTION 6:" + cClr + breakspace )
			// ========== ========== ========== ========== ==========

			fmt.Print( breakspace )
			fmt.Print( breakline )
			fmt.Print( breakspace )
			fmt.Print( "  [" + cGreen + "enter" + cClr + "] " + cCyan + "Return To Menu" + cClr + breakspace + "      [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()

		} else {

			// ========== ========== ========== ========== ==========
			fmt.Print(breakline)
			fmt.Print( breakspace + cBold + cCyan + "  Application Menu:" + cClr + breakspace + breakspace )

			// Webserver Status
			webserverStatus := ``
			if ( webserverRunning ) {
				webserverStatus = cBold + cGreen + "RUNNING" + cClr
			} else {
				webserverStatus = cBold + cRed + "STOPPED" + cClr
			}

			fmt.Print(
				"  [" + cGreen + "0" + cClr + "] " + cCyan + "Golang Console Color Testing" + cClr + breakspace +
				"  [" + cGreen + "1" + cClr + "] " + cCyan + "Simple Addition: Add(x,y)" + cClr + breakspace +
				"  [" + cGreen + "2" + cClr + "] " + cCyan + "Add Array Values in Loop: SumArray( array )	" + cClr + breakspace +
				"  [" + cGreen + "3" + cClr + "] " + cCyan + "Load JSON into Array" + cClr + breakspace +
				"  [" + cGreen + "4" + cClr + "] " + cCyan + "Reorder Array Alphabetically" + cClr + breakspace +
				"  [" + cGreen + "5" + cClr + "] " + cCyan + "Goroutines Communicating via Channels" + cClr + breakspace +
				"  [" + cGreen + "6" + cClr + "] " + cCyan + "Option 6" + cClr + breakspace +

				"  [" + cRed + "s" + cClr + "] " + cCyan + "Start/Stop Internal Webserver " + cClr + "[" + webserverStatus + "]" + cClr + breakspace +
				"  [" + cRed + "t" + cClr + "] " + cCyan + "Run Golang Tests" + cClr + "              " + "\\ http://localhost:9090" + breakspace + // Unicode: ↪
				"  [" + cRed + "q" + cClr + "] " + cCyan + "Quit Application" + cClr + breakspace )


			fmt.Print(breakline)
			fmt.Print( breakspace )
			fmt.Print( cBold + cCyan + "  Enter a Numeric Selection: " + cClr + breakspace + cYellow + "  > " + cClr )
			scanner.Scan()
			inputText = scanner.Text()


			if inputText != "q" {
				//fmt.Println("Your text was: ", inputText)
			}

			// selectedOption := 0

			// Array of options, if i = selectedOption, draw carrot in front ">"

			// fmt.Print(breakline)
			// ========== ========== ========== ========== ==========

		}
	}
}
